File: python_code/comm_utils.py
import json
import socket

import global_variables as gv

HEADER_LENGTH = gv.HEADER_LENGTH
FORMAT = gv.FORMAT

def rec_msg(client_socket: socket.socket):
    """
    Description: This function receives a message from a client socket.
    The message type is json and the message is received in two parts:
    1. The length of the message
    2. The message itself
    json is turned into a dictionary and returned.
    """
    len_msg = b""
    while True:
        msg_part = client_socket.recv(HEADER_LENGTH)
        if not msg_part:
            break
        len_msg += msg_part
        if len(len_msg) == HEADER_LENGTH:
            break
    
    msg_len = int(len_msg.decode(FORMAT).strip())

    msg = b""
    while True:
        msg_part = client_socket.recv(msg_len)
        if not msg_part:
            break
        msg += msg_part
        if len(msg) == msg_len:
            break
    
    msg = msg.decode(FORMAT)
    msg = json.loads(msg)
    return msg

def send_msg(client_socket: socket.socket, msg):
    """
    Description: This function sends a message to a client socket.
    The message type is json and the message is sent in two parts:
    1. The length of the message
    2. The message itself
    """
    try:
        msg = json.dumps(msg).encode(FORMAT)
        send_msg = f"{len(msg):<{HEADER_LENGTH}}".encode(FORMAT) + msg
        client_socket.sendall(send_msg)
    except Exception as e:
        print(f"[ERROR] sending message: {e}")

File: java_code/edcs/src/main/edcs/project/CommUtils.java
package main.edcs.project;

import java.io.*;
import java.net.Socket;
import java.nio.charset.StandardCharsets;

import org.json.JSONObject;

public class CommUtils {
    private static final int HEADER_LENGTH = GlobalVariables.HEADER_LENGTH;

    /*
    Description: This function receives a message from a client socket.
    The message type is json and the message is received in two parts:
    1. The length of the message
    2. The message itself
    the message is turned into a JSONObject and returned.
    * */
    public static JSONObject recMsg(Socket clientSocket) throws EOFException, IOException {
        InputStream inputStream = clientSocket.getInputStream();
        DataInputStream dataInputStream = new DataInputStream(inputStream);
        byte[] lenMsgBytes = new byte[HEADER_LENGTH];
        dataInputStream.readFully(lenMsgBytes, 0, HEADER_LENGTH);

        String d = new String(lenMsgBytes, StandardCharsets.UTF_8).trim();
        int msgLen = Integer.parseInt(d);
        byte[] msgBytes = new byte[msgLen];
        dataInputStream.readFully(msgBytes, 0, msgLen);

        String msgString = new String(msgBytes, StandardCharsets.UTF_8);

        return new JSONObject(msgString);
    }

    /*
    Description: This function sends a message to a client socket.
    The message type is JSONObject and the message is sent in two parts:
    1. The length of the message
    2. The message itself
    * */
    public static void sendMsg(Socket clientSocket, JSONObject msg) throws IOException {
        OutputStream outputStream = clientSocket.getOutputStream();
        String msgString = msg.toString();
        byte[] msgBytes = msgString.getBytes(StandardCharsets.UTF_8);
        String header = String.format("%-" + HEADER_LENGTH + "s", msgBytes.length);
        byte[] headerBytes = header.getBytes(StandardCharsets.UTF_8);

        outputStream.write(headerBytes, 0, headerBytes.length);
        outputStream.write(msgBytes, 0, msgBytes.length);
        outputStream.flush();
    }
}
